[gardena] Binding runs into 403 errors after some time

After seeing similar issues as [this forum user](https://community.openhab.org/t/openhab-3-4-release-discussion/142257/26), I recently recreated my OH application at the Husqvarna developer site. After doing this, the Gardena binding was working fine for some time, just recreating the Websocket every 2 hours:

```
2023-01-06 14:33:58.598 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-06 16:34:02.792 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-06 18:34:06.619 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-06 20:34:10.429 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-06 22:34:14.212 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
```

However, for unknown reasons, it started refreshing the Websocket every minute afterwards:
```
2023-01-06 22:34:14.212 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-06 23:55:48.051 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-06 23:56:01.067 [WARN ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice failed (a951c3474f-My Garden): java.net.UnknownHostException: api.smart.gardena.dev: Temporärer Fehler bei der Namensauflösung, restarting binding
2023-01-07 01:58:03.399 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 03:11:37.236 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 03:11:41.105 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 03:12:44.776 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 03:13:48.576 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 03:14:52.360 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 03:15:56.214 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 03:17:00.076 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 03:18:03.909 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 03:19:07.747 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 03:20:11.593 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 03:21:15.431 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 03:22:19.187 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 03:23:23.155 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 03:24:27.009 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 03:25:30.878 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
```

It continued doing so until the API (seemingly?) cut off my access:
```
2023-01-07 14:57:57.032 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 14:59:00.887 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 15:00:04.738 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 15:01:08.623 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 15:02:12.427 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 15:03:16.236 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-07 15:03:19.434 [WARN ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice failed (a951c3474f-My Garden): Error 403 Forbidden, {"Message":"User is not authorized to access this resource with an explicit deny"}, restarting binding
2023-01-07 15:05:20.047 [WARN ] [ternal.handler.GardenaAccountHandler] - Error 403 Forbidden, {"Message":"User is not authorized to access this resource with an explicit deny"}
2023-01-07 15:20:30.588 [WARN ] [ternal.handler.GardenaAccountHandler] - Error 403 Forbidden, {"Message":"User is not authorized to access this resource with an explicit deny"}
2023-01-07 15:35:41.188 [WARN ] [ternal.handler.GardenaAccountHandler] - Error 403 Forbidden, {"Message":"User is not authorized to access this resource with an explicit deny"}
2023-01-07 15:50:51.861 [WARN ] [ternal.handler.GardenaAccountHandler] - Error 403 Forbidden, {"Message":"User is not authorized to access this resource with an explicit deny"}
```

... so there seems to be an issue somewhere around failure handling. I'll recreate the API access and set the binding logging to TRACE to see whether I can gather more information on why this is happening.

## Your Environment
* Version used: 3.4
* Environment name and version: `OpenJDK 64-Bit Server VM (build 11.0.17+8-post-Ubuntu-1ubuntu220.04, mixed mode, sharing)`
* Operating System and version: Ubuntu 20.04 server

```2023-01-11 03:46:22.888 [TRACE] [rdena.internal.GardenaSmartWebSocket] - Sending ping (a951c3474f-My Garden)
2023-01-11 03:46:22.940 [TRACE] [rdena.internal.GardenaSmartWebSocket] - Pong received (a951c3474f-My Garden)
2023-01-11 03:48:22.889 [TRACE] [rdena.internal.GardenaSmartWebSocket] - Sending ping (a951c3474f-My Garden)
2023-01-11 03:48:22.941 [TRACE] [rdena.internal.GardenaSmartWebSocket] - Pong received (a951c3474f-My Garden)
2023-01-11 03:50:22.890 [TRACE] [rdena.internal.GardenaSmartWebSocket] - Sending ping (a951c3474f-My Garden)
2023-01-11 03:50:22.959 [TRACE] [rdena.internal.GardenaSmartWebSocket] - Pong received (a951c3474f-My Garden)
2023-01-11 03:52:22.891 [TRACE] [rdena.internal.GardenaSmartWebSocket] - Sending ping (a951c3474f-My Garden)
2023-01-11 03:52:22.943 [TRACE] [rdena.internal.GardenaSmartWebSocket] - Pong received (a951c3474f-My Garden)
2023-01-11 03:54:22.892 [TRACE] [rdena.internal.GardenaSmartWebSocket] - Sending ping (a951c3474f-My Garden)
2023-01-11 03:54:22.945 [TRACE] [rdena.internal.GardenaSmartWebSocket] - Pong received (a951c3474f-My Garden)
2023-01-11 03:56:22.893 [TRACE] [rdena.internal.GardenaSmartWebSocket] - Sending ping (a951c3474f-My Garden)
2023-01-11 03:56:22.946 [TRACE] [rdena.internal.GardenaSmartWebSocket] - Pong received (a951c3474f-My Garden)
2023-01-11 03:58:22.894 [TRACE] [rdena.internal.GardenaSmartWebSocket] - Sending ping (a951c3474f-My Garden)
2023-01-11 04:00:22.895 [TRACE] [rdena.internal.GardenaSmartWebSocket] - Sending ping (a951c3474f-My Garden)
2023-01-11 04:04:52.897 [DEBUG] [rdena.internal.GardenaSmartWebSocket] - Gardena Webservice error (a951c3474f-My Garden)
org.eclipse.jetty.websocket.api.CloseException: java.util.concurrent.TimeoutException: Idle timeout expired: 150000/150000 ms
        at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.onReadTimeout(AbstractWebSocketConnection.java:564) ~[?:?]
        at org.eclipse.jetty.io.AbstractConnection.onFillInterestedFailed(AbstractConnection.java:172) ~[?:?]
        at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.onFillInterestedFailed(AbstractWebSocketConnection.java:539) ~[?:?]
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.failed(AbstractConnection.java:317) ~[?:?]
        at org.eclipse.jetty.io.FillInterest.onFail(FillInterest.java:140) ~[?:?]
        at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.onFillableFail(SslConnection.java:591) ~[?:?]
        at org.eclipse.jetty.io.ssl.SslConnection.onFillInterestedFailed(SslConnection.java:419) ~[?:?]
        at org.eclipse.jetty.io.ssl.SslConnection$2.failed(SslConnection.java:170) ~[?:?]
        at org.eclipse.jetty.io.FillInterest.onFail(FillInterest.java:140) ~[?:?]
        at org.eclipse.jetty.io.AbstractEndPoint.onIdleExpired(AbstractEndPoint.java:407) ~[?:?]
        at org.eclipse.jetty.io.IdleTimeout.checkIdleTimeout(IdleTimeout.java:171) ~[?:?]
        at org.eclipse.jetty.io.IdleTimeout.idleCheck(IdleTimeout.java:113) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: java.util.concurrent.TimeoutException: Idle timeout expired: 150000/150000 ms
        ... 8 more
2023-01-11 04:04:52.910 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-11 04:04:52.911 [DEBUG] [rdena.internal.GardenaSmartWebSocket] - Closing Gardena Webservice (a951c3474f-My Garden)
2023-01-11 04:04:55.911 [TRACE] [ng.gardena.internal.GardenaSmartImpl] - >>> POST https://api.smart.gardena.dev/v1/websocket, data REDACTED
2023-01-11 04:04:55.912 [DEBUG] [ng.gardena.internal.GardenaSmartImpl] - Gardena API token valid
2023-01-11 04:04:55.912 [DEBUG] [ng.gardena.internal.GardenaSmartImpl] - Token expiration: accessToken: 2023-01-11T17:17:04.940233+01:00[Europe/Berlin], refreshToken: 2023-01-20T17:16:15.940093+01:00[Europe/Berlin]
2023-01-11 04:04:56.474 [TRACE] [ng.gardena.internal.GardenaSmartImpl] - <<< status:201, {"data": REDACTED}
2023-01-11 04:04:56.474 [DEBUG] [rdena.internal.GardenaSmartWebSocket] - Reconnecting to Gardena Webservice (a951c3474f-My Garden)
2023-01-11 04:04:56.811 [DEBUG] [rdena.internal.GardenaSmartWebSocket] - Connected to Gardena Webservice (a951c3474f-My Garden)
2023-01-11 04:04:56.811 [DEBUG] [rdena.internal.GardenaSmartWebSocket] - Connection to Gardena Webservice was closed (a951c3474f-My Garden): code: 1001, reason: java.util.concurrent.TimeoutException: Idle timeout expired: 150000/150000 ms
2023-01-11 04:04:56.812 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-11 04:04:56.812 [DEBUG] [rdena.internal.GardenaSmartWebSocket] - Closing Gardena Webservice (a951c3474f-My Garden)
2023-01-11 04:04:56.866 [DEBUG] [rdena.internal.GardenaSmartWebSocket] - Connection to Gardena Webservice was closed (a951c3474f-My Garden): code: 1000, reason: null
2023-01-11 04:04:59.813 [TRACE] [ng.gardena.internal.GardenaSmartImpl] - >>> POST https://api.smart.gardena.dev/v1/websocket, data: REDACTED
2023-01-11 04:04:59.813 [DEBUG] [ng.gardena.internal.GardenaSmartImpl] - Gardena API token valid
2023-01-11 04:04:59.814 [DEBUG] [ng.gardena.internal.GardenaSmartImpl] - Token expiration: accessToken: 2023-01-11T17:17:04.940233+01:00[Europe/Berlin], refreshToken: 2023-01-20T17:16:15.940093+01:00[Europe/Berlin]
2023-01-11 04:05:00.239 [TRACE] [ng.gardena.internal.GardenaSmartImpl] - <<< status:201, {"data": REDACTED}
2023-01-11 04:05:00.240 [DEBUG] [rdena.internal.GardenaSmartWebSocket] - Reconnecting to Gardena Webservice (a951c3474f-My Garden)
2023-01-11 04:05:00.461 [DEBUG] [rdena.internal.GardenaSmartWebSocket] - Connected to Gardena Webservice (a951c3474f-My Garden)
2023-01-11 04:05:00.819 [TRACE] [rdena.internal.GardenaSmartWebSocket] - <<< event (a951c3474f-My Garden): REDACTED
2023-01-11 04:05:00.832 [TRACE] [rdena.internal.GardenaSmartWebSocket] - <<< event (a951c3474f-My Garden): REDACTED
2023-01-11 04:05:00.842 [TRACE] [rdena.internal.GardenaSmartWebSocket] - <<< event (a951c3474f-My Garden): REDACTED
2023-01-11 04:05:00.854 [TRACE] [rdena.internal.GardenaSmartWebSocket] - <<< event (a951c3474f-My Garden): REDACTED
2023-01-11 04:06:00.514 [DEBUG] [rdena.internal.GardenaSmartWebSocket] - Connection to Gardena Webservice was closed (a951c3474f-My Garden): code: 1000, reason: Timeout manually closing dead connection (a951c3474f-My Garden)
2023-01-11 04:06:00.515 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-11 04:06:00.515 [DEBUG] [rdena.internal.GardenaSmartWebSocket] - Closing Gardena Webservice (a951c3474f-My Garden)
2023-01-11 04:06:03.516 [TRACE] [ng.gardena.internal.GardenaSmartImpl] - >>> POST https://api.smart.gardena.dev/v1/websocket, data: REDACTED
2023-01-11 04:06:03.516 [DEBUG] [ng.gardena.internal.GardenaSmartImpl] - Gardena API token valid
2023-01-11 04:06:03.517 [DEBUG] [ng.gardena.internal.GardenaSmartImpl] - Token expiration: accessToken: 2023-01-11T17:17:04.940233+01:00[Europe/Berlin], refreshToken: 2023-01-20T17:16:15.940093+01:00[Europe/Berlin]
2023-01-11 04:06:04.228 [TRACE] [ng.gardena.internal.GardenaSmartImpl] - <<< status:201, {"data": REDACTED}
2023-01-11 04:06:04.228 [DEBUG] [rdena.internal.GardenaSmartWebSocket] - Reconnecting to Gardena Webservice (a951c3474f-My Garden)
2023-01-11 04:06:04.553 [DEBUG] [rdena.internal.GardenaSmartWebSocket] - Connected to Gardena Webservice (a951c3474f-My Garden)
2023-01-11 04:06:04.854 [TRACE] [rdena.internal.GardenaSmartWebSocket] - <<< event (a951c3474f-My Garden): REDACTED
2023-01-11 04:06:04.865 [TRACE] [rdena.internal.GardenaSmartWebSocket] - <<< event (a951c3474f-My Garden): REDACTED
2023-01-11 04:06:04.874 [TRACE] [rdena.internal.GardenaSmartWebSocket] - <<< event (a951c3474f-My Garden): REDACTED
2023-01-11 04:06:04.884 [TRACE] [rdena.internal.GardenaSmartWebSocket] - <<< event (a951c3474f-My Garden): REDACTED
2023-01-11 04:07:04.607 [DEBUG] [rdena.internal.GardenaSmartWebSocket] - Connection to Gardena Webservice was closed (a951c3474f-My Garden): code: 1000, reason: Timeout manually closing dead connection (a951c3474f-My Garden)
2023-01-11 04:07:04.608 [INFO ] [ng.gardena.internal.GardenaSmartImpl] - Restarting GardenaSmart Webservice (a951c3474f-My Garden)
2023-01-11 04:07:04.608 [DEBUG] [rdena.internal.GardenaSmartWebSocket] - Closing Gardena Webservice (a951c3474f-My Garden)
```

This is the start of the minutely reconnections. From there on the reconnection loop continues to run every minute, looking exactly the same each time.

--- a/bundles/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/internal/GardenaSmartWebSocket.java
+++ b/bundles/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/internal/GardenaSmartWebSocket.java
@@ -16,7 +16,6 @@ import java.io.IOException;
 import java.net.URI;
 import java.nio.ByteBuffer;
 import java.nio.charset.StandardCharsets;
-import java.time.Instant;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.TimeUnit;
@@ -49,12 +48,12 @@ import org.slf4j.LoggerFactory;
 public class GardenaSmartWebSocket {
     private final Logger logger = LoggerFactory.getLogger(GardenaSmartWebSocket.class);
     private final GardenaSmartWebSocketListener socketEventListener;
-    private final long WEBSOCKET_IDLE_TIMEOUT = 300;
+    private final int MAX_UNANSWERED_PINGS = 5;
 
     private WebSocketSession session;
     private WebSocketClient webSocketClient;
     private boolean closing;
-    private Instant lastPong = Instant.now();
+    private int unansweredPings = 0;
     private ScheduledExecutorService scheduler;
     private @Nullable ScheduledFuture<?> connectionTracker;
     private ByteBuffer pingPayload = ByteBuffer.wrap("ping".getBytes(StandardCharsets.UTF_8));
@@ -115,8 +114,9 @@ public class GardenaSmartWebSocket {
     }
 
     @OnWebSocketConnect
-    public void onConnect(Session session) {
+    public synchronized void onConnect(Session session) {
         closing = false;
+        unansweredPings = 0;
         logger.debug("Connected to Gardena Webservice ({})", socketId);
 
         ScheduledFuture<?> connectionTracker = this.connectionTracker;
@@ -129,9 +129,9 @@ public class GardenaSmartWebSocket {
     }
 
     @OnWebSocketFrame
-    public void onFrame(Frame pong) {
+    public synchronized void onFrame(Frame pong) {
         if (pong instanceof PongFrame) {
-            lastPong = Instant.now();
+            unansweredPings = 0;
             logger.trace("Pong received ({})", socketId);
         }
     }
@@ -170,14 +170,14 @@ public class GardenaSmartWebSocket {
      */
     private synchronized void sendKeepAlivePing() {
         final PostOAuth2Response accessToken = token;
-        if ((Instant.now().getEpochSecond() - lastPong.getEpochSecond() > WEBSOCKET_IDLE_TIMEOUT) || accessToken == null
-                || accessToken.isAccessTokenExpired()) {
+        if (unansweredPings > MAX_UNANSWERED_PINGS || accessToken == null || accessToken.isAccessTokenExpired()) {
             session.close(1000, "Timeout manually closing dead connection (" + socketId + ")");
         } else {
             if (session.isOpen()) {
                 try {
                     logger.trace("Sending ping ({})", socketId);
                     session.getRemote().sendPing(pingPayload);
+                    ++unansweredPings;
                 } catch (IOException ex) {
                     logger.debug("Error while sending ping: {}", ex.getMessage());
                 }

